services/src/kilda-pce/src/main/java/org/openkilda/pce/impl/BestCostAndShortestPathFinder.java

```diff
@@ -0,0 +1,301 @@
+                    format("SOURCE node doesn't exist. It isn't in the AVAILABLE network: %s", startSwitchId));
+        }
+
+        Switch end = network.getSwitch(endSwitchId);
+        if (end == null) {
+            throw new SwitchNotFoundException(
+                    format("DESTINATION node doesn't exist. It isn't in the AVAILABLE network: %s", endSwitchId));
+        }
+
+        List<Isl> forwardPath = getPath(start, end, Integer.MAX_VALUE);
+        List<Isl> reversePath = getPath(end, start, Integer.MAX_VALUE, forwardPath);
+        return Pair.of(forwardPath, reversePath);
+    }
+
+    /**
+     * Call this method to find a path from start to end (srcDpid to dstDpid), particularly if you have no idea if the
+     * path exists or what the best path is.
+     *
+     * @return An ordered list that represents the path from start to end, or an empty list
+     */
+    private List<Isl> getPath(Switch start, Switch end, long maximumCost) {
+        long bestCost = maximumCost; // Need to be long because it stores sum of ints.
+        SearchNode bestPath = null;
+
+        Deque<SearchNode> toVisit = new LinkedList<>(); // working list
+        Map<Switch, SearchNode> visited = new HashMap<>();
+
+        toVisit.add(new SearchNode(start, allowedDepth, 0, emptyList()));
+
+        while (!toVisit.isEmpty()) {
+            SearchNode current = toVisit.pop();
+
+            // Determine if this node is the destination node.
+            if (current.dstSw.equals(end)) {
+                // We found the destination
+                if (current.parentCost < bestCost) {
+                    // We found a best path. If we don't get here, then the entire graph will be
+                    // searched until we run out of nodes or the depth is reached.
+                    bestCost = current.parentCost;
+                    bestPath = current;
+                }
+                // We found dest, no need to keep processing
+                continue;
+            }
+
+            // Otherwise, if we've been here before, see if this path is better
+            SearchNode prior = visited.get(current.dstSw);
+            if (prior != null && current.parentCost >= prior.parentCost) {
+                continue;
+            }
+
+            // Stop processing entirely if we've gone too far, or over bestCost
+            if (current.allowedDepth <= 0 || current.parentCost > bestCost) {
+                continue;
+            }
+
+            // Either this is the first time, or this is cheaper .. either way, this node should
+            // be the one in the visited list
+            visited.put(current.dstSw, current);
+
+
+            // At this stage .. haven't found END, haven't gone too deep, and we are not over cost.
+            // So, add the outbound isls.
+            current.dstSw.getOutgoingLinks().stream()
+                    .sorted(Comparator.comparing(isl -> isl.getDestSwitch().getSwitchId()))
```

Comment:
Why are you sorting ISLs here? As I understand you make depth search and sorting here doesn't help much. What about using PriorityQueue for toVisit instead on LinkedList?